### DISCUSSION TITLE
Einrichtung Zigbee-Steckdose mit Energiemessung

### DISCUSSION THREAD

**Discussion by bismosa** [camueller/SmartApplianceEnabler#607] [Q&A]
Hallo!

Ich starte erst mit dem SAE. Meine PV-Anlage ist auch erst seit wenigen Wochen in Betrieb.  Vielen Dank für die Tolle Arbeit.
Ich möchte als erstes Gerät eine Zigbee-Steckdose mit Energiemessung integrieren. 

Die ersten Sachen funktionieren schon. Durch das Portal wird die Steckdose bei PV-Überschuss ein- und ausgeschaltet.

1.)
Ich bekomme als Statusmeldung von der Steckdose:
`MQTT message received: {"child_lock":"UNLOCK","current":0,"energy":61.07,"indicator_mode":"off/on","last_seen":"2024-05-22T10:10:16.732Z","linkquality":83,"power":0,"power_outage_memory":"on","state":"OFF","voltage":229}`

Wird die Steckdose manuell geschaltet, wird der Status auch (nach einiger Zeit) übernommen. Sicher bin ich mir aber nicht, ob ich das richtig eingestellt habe.
![image](https://github.com/camueller/SmartApplianceEnabler/assets/2944275/0b481857-12ba-4471-a2fa-2574f0a8f702)
Stimmt mein Regex hier?
Müsste ich nicht eher sowas machen wie:
![image](https://github.com/camueller/SmartApplianceEnabler/assets/2944275/769c4f0d-4308-4a47-b71b-716a78dcefef)
Aber wonach wird ein "on" oder "off" ausgewertet? Ob der Wert vorhanden ist? Oder true/false?

2.)
Beim Zähler ist auf jeden Fall noch etwas falsch:
![image](https://github.com/camueller/SmartApplianceEnabler/assets/2944275/7672cbb9-5812-49da-a82d-a059b6ac48a8)
So habe ich es in der Doku verstanden. Die Log Datei sagt aber:
```
ERROR [MQTT Call: F-10071986-000000000001-00-MqttElectricityMeter-16] d.a.s.m.MqttClient [MqttClient.java:175] F-10071986-000000000001-00-MQTT-MqttElectricityMeter: MQTT connecton lost {}
org.eclipse.paho.client.mqttv3.MqttException: MqttException
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:228)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: path can not be null or empty
	at com.jayway.jsonpath.internal.Utils.notEmpty(Utils.java:383)
	at com.jayway.jsonpath.internal.JsonContext.read(JsonContext.java:76)
	at de.avanux.smartapplianceenabler.protocol.JsonContentProtocolHandler.readValue(JsonContentProtocolHandler.java:39)
	at de.avanux.smartapplianceenabler.meter.MqttElectricityMeter.lambda$start$0(MqttElectricityMeter.java:160)
	at de.avanux.smartapplianceenabler.mqtt.MqttClient.lambda$subscribeCachedMessageHandler$1(MqttClient.java:330)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.deliverMessage(CommsCallback.java:511)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.handleMessage(CommsCallback.java:417)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:214)
	... 1 common frames omitted
```
Ich habe es auch inunterschiedlichen Regex-Varianten probiert. Aber hier lag ich auch meist falsch.
Z.B.:
`(?<=\"power\":)[^\,.]*`
Hiermit bekomme ich dann keine Fehlermeldungen mehr:
`,\"power\":((\d*\.)?\d+)`
Aber im Portal wird der Verbrauch immer mit 500W (max. Leistungsaufnahme) angenommen. 
Irgendwo hatte ich gelesen, das man bei min. Leistungsaufnahme auch etwas eintragen muss damit es funktioniert. Hier habe ich jetzt "0" eingetragen.
![image](https://github.com/camueller/SmartApplianceEnabler/assets/2944275/54aeb070-d835-4e3c-8377-f2927e9fb064)
Leider auch ohne Erfolg...

3.)
Welchen Unterschied gibt es eigentlich bei den unterschiedlichen Gerätetypen? Gibt es da eine Übersicht?

Danke :)

**Comment by camueller** [comment]
zu 1) Steht in der Doku: "Für das Status-Topic muss als Regex für Status-Extraktion ein [regulärer Ausdruck](https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/ValueExtraction_DE.md) angegeben werden der "matchen" muss, wenn der Schaltzustand "eingeschaltet" ist. Liefert dieser beispielsweise "ON" (ohne Anführungszeichen), so muss als regulärer Ausdruck "(ON)" (ohne Anführungszeichen) eingetragen werden."

zu 2) Ich würde zunächst einmal die Probleme mit der MQTT-Verbindung beheben: 
"MQTT connecton lost"

zu 3) Die Liste ist von SMA und für den SAE ist der Typ egal bis auf "Elektroauto-Ladegerät" - dieses hat diverse Besonderheiten.

**Comment by bismosa** [comment]
Danke für die Rückmeldung!

1.) Alles klar. Habe ich nun auch gefunden und verstanden.  Also ist mein Regex:
`("state":"ON")`

2.) Die MQTT-Verbindung bricht nur ab, wenn die Einstellung im Pfad für Wert-Extraktion oder im Regex falsch ist.
Mit einem Pfad habe ich es nicht hinbekommen. Mit diesem Regex schon:
`,\"power\":((\d*\.)?\d+)`

Mittlerweile wird dann auch im Portal die Leistung richtig angezeigt. 

3.) Danke für die Info :)

Aller Anfang ist schwer. Mir hätten weitere Beispiele vermutlich gut geholfen. Vielleicht wäre es praktisch, wenn man Konfigbeispiele aus dem UI direkt in einem lesbaren Format exportieren (anonymisiert) könnte. So sieht die Konfig jetzt bei mir für einen Zigbee-Schalter aus, der nur bei PV-Überschuss eingeschaltet werden soll (und so lange wie Möglich in Betrieb sein soll):

```
<Appliance id="F-12345678-000000000001-00">
	<MqttSwitch topic="zigbee2mqtt/0xa4c138b803e99999/set" onPayload="ON" offPayload="OFF" statusTopic="zigbee2mqtt/0xa4c138b803e99999" statusExtractionRegex="(&quot;state&quot;:&quot;ON&quot;)"/>
	<MqttElectricityMeter topic="zigbee2mqtt/0xa4c138b803e99999" name="Power" extractionRegex=",\&quot;power\&quot;:((\d*\.)?\d+)" factorToValue="1.0"/>
	<Schedule enabled="true">
		<RuntimeRequest min="0" max="86160" enabled="true"/>
		<DayTimeframe>
			<Start hour="0" minute="0" second="0"/>
			<End hour="23" minute="58" second="0"/>
			<DayOfWeek>1</DayOfWeek>
			<DayOfWeek>2</DayOfWeek>
			<DayOfWeek>3</DayOfWeek>
			<DayOfWeek>4</DayOfWeek>
			<DayOfWeek>5</DayOfWeek>
			<DayOfWeek>6</DayOfWeek>
			<DayOfWeek>7</DayOfWeek>
			<DayOfWeek>8</DayOfWeek>
		</DayTimeframe>
	</Schedule>
</Appliance>
```
Wobei 
statusExtractionRegex = ("state":"ON")
extractionRegex = ,\"power\":((\d*\.)?\d+)

Bisher scheint das so zu funktionieren.
